xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/XWikiServletURLFactory.java

```diff
@@ -613,6 +613,39 @@ public URL createAttachmentURL(String filename, String spaces, String name, Stri
             }
         }
 
+        if (isContextDoc(xwikidb, spaces, name, context)) {
+            XWikiAttachment attachment = context.getDoc().getAttachment(filename);
+            if (attachment != null) {
+                return createAttachmentRevisionURL(filename, spaces, name, attachment.getVersion(),
+                    querystring, xwikidb, context);
+            } else {
+                if (LOGGER.isErrorEnabled()) {
+                    LOGGER.error(String.format("Exception while trying to get attachment [%s] from [%s.%s]!",
+                        filename, spaces, name));
```

Comment:
Best practice: always use SLF4J syntax :)

```
LOGGER.error("Exception while trying to get attachment [{}] from space [{}], page [{}]!", filename, spaces, name));
```

2nd best practice: always separate each variable with `[]`. Imagine case 1 (spaces ="Hello.World" and name = "") and case 2 (spaces="" and name "Hello.World"), you wouldn't be able to differentiate them when debugging a production instance.

Reply:
Best practices at https://dev.xwiki.org/xwiki/bin/view/Community/CodeStyle/JavaCodeStyle/#HLoggingBestPractices